### DISCUSSION TITLE
Introducing TorchSparseGradUtils - v0.1.0

### DISCUSSION THREAD

**Discussion by theo-barfoot** [cai4cai/torchsparsegradutils#46] [Announcements]
# Release Notes:

We are excited to announce the first release of TorchSparseGradUtils, a suite of efficient utilities that extend the functionality of PyTorch sparse tensor operations to support sparse gradient back-propagation from sparse input tensors.

Here are the key features included in this release:

## PyTorch Matrix operations with sparse gradients support:

1. Sparse-Dense matrix multiplication with batch support [sparse_matmul](https://github.com/cai4cai/torchsparsegradutils/blob/4bbddbb8a38d278f9b6582e3ddfd01119616ac85/torchsparsegradutils/sparse_matmul.py#L5C1-L5C1).
2. Sparse-Dense triangular linear solver with batch support [sparse_triangular_solve](https://github.com/cai4cai/torchsparsegradutils/blob/4bbddbb8a38d278f9b6582e3ddfd01119616ac85/torchsparsegradutils/sparse_solve.py#L5)
3. Sparse-Dense generic linear solver [sparse_generic_solve](https://github.com/cai4cai/torchsparsegradutils/blob/4bbddbb8a38d278f9b6582e3ddfd01119616ac85/torchsparsegradutils/sparse_solve.py#L223)

## Sparse Gaussian Distribution:

1. [SparseMultivariateNormal](https://github.com/cai4cai/torchsparsegradutils/blob/4bbddbb8a38d278f9b6582e3ddfd01119616ac85/torchsparsegradutils/distributions/sparse_multivariate_normal.py#L48C7-L48C31) Distribution parameterised by either a sparse lower triangular covariance or precision matrix with reparameterised sampling

## Sparse Encoder:

1. [PairwiseVoxelEncoder](https://github.com/cai4cai/torchsparsegradutils/blob/4bbddbb8a38d278f9b6582e3ddfd01119616ac85/torchsparsegradutils/encoders/pairwise_voxel_encoder.py#L260) to encode sparse matrices with relationships between pairs of voxels in local neighbourshoods of 3D volumetric images.

## Sparse utilities:

1. Convert COO indices and values to CSR indices and values with [convert_coo_to_csr_indices_values](https://github.com/cai4cai/torchsparsegradutils/blob/4bbddbb8a38d278f9b6582e3ddfd01119616ac85/torchsparsegradutils/utils/utils.py#L81C5-L81C38), with batch support.
2. Convert COO sparse tensors to CSR sparse tensors with batched support with [convert_coo_to_csr](https://github.com/cai4cai/torchsparsegradutils/blob/4bbddbb8a38d278f9b6582e3ddfd01119616ac85/torchsparsegradutils/utils/utils.py#L146C5-L146C23)
3. Equivalent of torch.block_diag() for sparse COO and CSR matrices [sparse_block_diag](https://github.com/cai4cai/torchsparsegradutils/blob/4bbddbb8a38d278f9b6582e3ddfd01119616ac85/torchsparsegradutils/utils/utils.py#L189C5-L189C22) and a function to perform the reverse [sparse_block_diag_split](https://github.com/cai4cai/torchsparsegradutils/blob/4bbddbb8a38d278f9b6582e3ddfd01119616ac85/torchsparsegradutils/utils/utils.py#L303C5-L303C28)
4. Equivalent of torch.eye for sparse COO and CSR matrices [sparse_eye](https://github.com/cai4cai/torchsparsegradutils/blob/4bbddbb8a38d278f9b6582e3ddfd01119616ac85/torchsparsegradutils/utils/utils.py#L403C5-L403C15)
5. Equivalent of torch.stack() for sparse CSR tensors [stack_csr](https://github.com/cai4cai/torchsparsegradutils/blob/4bbddbb8a38d278f9b6582e3ddfd01119616ac85/torchsparsegradutils/utils/utils.py#L4C5-L4C14)

## Generating Random Sparse Matrices:

1. Equivalent of torch.rand() for sparse COO and CSR matrices [rand_sparse](https://github.com/cai4cai/torchsparsegradutils/blob/4bbddbb8a38d278f9b6582e3ddfd01119616ac85/torchsparsegradutils/utils/random_sparse.py#L20C5-L20C16)
2. [rand_sparse_tri](https://github.com/cai4cai/torchsparsegradutils/blob/4bbddbb8a38d278f9b6582e3ddfd01119616ac85/torchsparsegradutils/utils/random_sparse.py#L41C5-L41C20) used for generating random strictly triangular sparse matrices in either COO or CSR format.

## Additional backbone solvers implemented in pure PyTorch:

1. BICGSTAB (ported from [pykrylov](https://github.com/PythonOptimizers/pykrylov)) [bicstab](https://github.com/cai4cai/torchsparsegradutils/blob/4bbddbb8a38d278f9b6582e3ddfd01119616ac85/torchsparsegradutils/utils/bicgstab.py#L25C8-L25C8)
6. CG (ported from [cornellius-gp/linear_operator](https://github.com/cornellius-gp/linear_operator)) [linear_cg](https://github.com/cai4cai/torchsparsegradutils/blob/4bbddbb8a38d278f9b6582e3ddfd01119616ac85/torchsparsegradutils/utils/linear_cg.py#L101C5-L101C14) 
7. LSMR (ported from [pytorch-minimize](https://github.com/rfeinman/pytorch-minimize)) [lsmr](https://github.com/cai4cai/torchsparsegradutils/blob/4bbddbb8a38d278f9b6582e3ddfd01119616ac85/torchsparsegradutils/utils/lsmr.py#L18)
8. MINRES (ported from [cornellius-gp/linear_operator](https://github.com/cornellius-gp/linear_operator)) [minres](https://github.com/cai4cai/torchsparsegradutils/blob/4bbddbb8a38d278f9b6582e3ddfd01119616ac85/torchsparsegradutils/utils/minres.py#L28C5-L28C11)

## CuPy and JAX solvers:

We also provide wrappers around [cupy sparse solvers](https://docs.cupy.dev/en/stable/reference/scipy_sparse_linalg.html#solving-linear-problems) and [jax sparse solvers](https://jax.readthedocs.io/en/latest/jax.scipy.html#module-jax.scipy.sparse.linalg). Allowing linear systems of PyTorch sparse matrices to be solved using a CuPy or JAX back-end:

1. Sparse-Dense linear solver with CuPy back-end [sparse_solve_c4t](https://github.com/cai4cai/torchsparsegradutils/blob/4bbddbb8a38d278f9b6582e3ddfd01119616ac85/torchsparsegradutils/cupy/cupy_sparse_solve.py#L5)
3. Sparse-Dense linear solver with JAX back-end [sparse_solve_j4t](https://github.com/cai4cai/torchsparsegradutils/blob/4bbddbb8a38d278f9b6582e3ddfd01119616ac85/torchsparsegradutils/jax/jax_sparse_solve.py#L13C5-L13C21)

## Installation:

This version can be installed using:

`pip install torchsparsegradutils==0.1.0`

We welcome any feedback, suggestions, and contributions via our [issues page](https://github.com/cai4cai/torchsparsegradutils/issues).

For more details about this release, you can refer to the [Full Changelog](https://github.com/cai4cai/torchsparsegradutils/commits/v0.1.0).

<hr /><em>This discussion was created from the release <a href='https://github.com/cai4cai/torchsparsegradutils/releases/tag/'>Introducing TorchSparseGradUtils - v0.1.0</a>.</em>
